client/util/Breadcrumbs.tsx

```diff
@@ -0,0 +1,87 @@
+export interface TreeNode {
+    pid: string;
+    title: string;
+    parents: Array<TreeNode>;
+}
+
+export interface TreeData {
+    topNodes: Array<string>;
+    records: Record<string, TreeNode>;
+    childLookups: Record<string, Array<string>>;
+}
+
+interface BreadcrumbTrail {
+    pid: string;
+    path: Array<TreeNode>;
+}
+
+/**
+ * Analyze the raw breadcrumb data, identifying top-level nodes (i.e. nodes with
+ * no parents) and creating lookup tables for children. This makes it easier to
+ * render breadcrumbs from left to right and to find all relevant trails.
+ */
+export function processBreadcrumbData(data: TreeNode): TreeData {
+    const queue: Array<TreeNode> = [data];
+    const topNodes: Array<string> = [];
+    const childLookups: Record<string, Array<string>> = {};
+    const records: Record<string, TreeNode> = {};
+    while (queue.length > 0) {
```

Comment:
breadth first search?

Reply:
You can maybe implement depth first search instead since array.shift() is actally O(n) for javascript instead of O(1).

Reply:
Good to know. This code is a direct port of an algorithm that I didn't write, so I'll need to spend some time analyzing it. That's probably a separate task, though, since this logic is already present in dev and has just been relocated for this PR.